**B03 Team Members:**   Himani Tyagi,    Jaishankar Govindaraj,    Leonardo Trucios Cevasco,    Srithijaa Sankepally,    Yumeng (Emelyn) Tang

# **Summary of the Assignment**

**Project title :** Global Pandemic And The Government - COVID-19’s Impact On Boston Government Employees’ Earnings

**Problem Definition:** Evaluate the Boston government sector to understand the impact of the global pandemic, COVID-19, on the employees’ earnings. 
The main topics that we will focus on covering are the followings: 

1. Examine the data across years preceding, during, and following the pandemic to assess whether there have been any effects on both earnings and the workforce.
2. Explore which job levels/positions have recorded the highest earnings, and which department have recorded the highest number of employees
3. Determine which income component has the most impact on an employee’s gross earnings.
4. Construct patterns from data collected across various Boston areas/zip codes.

**Data:** For the developement of this project we have used two datasets from two different sources.

1. Boston Government - Office of Human Resources: https://data.boston.gov/dataset/employee-earnings-report
  * The City of Boston releases an annual dataset containing payroll information for its employees, encompassing employee names, job details, and earnings breakdown, including base salary, overtime, and total compensation. To conduct a comprehensive analysis, we are using the dataset spanning the years 2018, 2019, 2020, 2021, and 2022. This resource, offers citizens valuable insights into the financial aspects of the City's workforce during these specific years.

2. United States Zip Codes - Massachusetts: https://www.unitedstateszipcodes.org/ma/
  * This dataset provides zip code information for Massachusetts, including key details such as zip code, type, primary city, state, county, timezone, latitude, longitude among others. It offers a concise and comprehensive snapshot of geographic and locational attributes associated with each zip code in the state of Massachusetts. By incorporating this dataset, the project can be analyzed in a location-wise level.


**Motivation :** Government jobs tend to be secure because they are essential for the functioning of society and are less influenced by market fluctuations compared to private sector. We aim to provide insights into Boston’s government sector for our peers with potential interest.

**Executive Summary:** This project assesses the effects of the COVID-19 pandemic on earnings in the Boston government sector. Utilizing datasets from 2018 to 2022, sourced from Boston Government's payroll records and Massachusetts zip code information, the study evaluates changes in workforce and earnings during the pre-pandemic, pandemic, and post-pandemic periods. Key objectives include identifying top-earning job levels, determining departments with the highest employee numbers, and analyzing income components impacting gross earnings. Emphasis is placed on constructing patterns across Boston areas and zip codes. Motivated by the stability of government jobs, our project aims to provide insights into Boston's government workforce over the last five years, which can be a source of interest for our peers.

# **Description of the Datasets**

## Description for Boston Government - Office of Human Resources Dataset

### Dictionary

*   **Name:** Full name of each government employee

*  **Department_Name:** The department in which the employee works

*   **Title:** The job title or position held by the employee

*   **Regular:** Regular earnings or base salary of the employee

*   **Retro:** Retroactive pay, if any, received by the employee. This concept refers to compensation added to an employee's paycheck to make up for a compensation shortfall in a previous pay period

*   **Other:** Additional earnings or extraordinary income for the employee. This type of compensation may consist of various supplementary forms of compensation, including bonuses, allowances, or reimbursement expenses

*   **Overtime:** Overtime payments received by the employee

*   **Injured:** Earnings related to injury, if applicable

*   **Detail:** Additional details payments or special job-related earnings

*   **Quinn Education:**  Earnings related to educational incentives or allowances

*   **Total Gross:** The total earnings of the employee, including regular, retro, other, overtime, injured, detail, and education earnings.

*   **Postal:** The zip code associated with the employee's location

*   **Year:** The year for which the earnings data is reported


### Basic Information for the concatenated dataset from 2018-2022 tables

* There are a total of 114523 rows
* There are a total of 12 columns
* Null values: 
    * Zip_code column returned 93 null values
    * Columns like regular, retro, other, overtime, injured, detail and education represent income components, hence, null values are taken as USD 0, which means that there has been no contribution towards that component

## United States Zip Codes - Massachusetts

### Dictionary

*   **zip:** The ZIP code, a numerical code used in postal addresses to facilitate mail delivery

*  **type:** The type of address associated with the ZIP code (e.g., STANDARD for standard delivery or PO BOX for a post office box address)

*   **primary_city:** The primary city or locality associated with the ZIP code

*   **state:**  The state in which the ZIP code is located (in this case, MA for Massachusetts)

*   **county:** The county to which the ZIP code belongs (e.g., Plymouth County)

*   **timezone:** The timezone of the area covered by the ZIP code (e.g., America/New_York)

*   **world_region:** The world region or continent to which the ZIP code is geographically associated (in this case, NA for North America)

*   **country:** The country to which the ZIP code belongs (in this case, US for the United States)

*   **latitude:** The geographic latitude coordinates of the ZIP code location

*   **longitude:** The geographic longitude coordinates of the ZIP code location

# **Entity Relationship Diagram**

In [ ]:
Paste the ERD updated

# **Exploratory Data Analysis**

Before doing all the data cleaning, we have eliminated the observations with similar names, so we can have 1 unique identifier for each of the 5 tables 
that represents the payroll of 2018,2019,2020,2021 and 2022

In [46]:
%%bigquery
CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.2022_unique`AS
SELECT *
FROM (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY NAME ORDER BY NAME) AS row_num
    FROM `ba-775-b03.BA_775_Dataset.2022` 
) AS subquery
WHERE row_num = 1;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

The table that has information of the payroll in 2022, has 23,204 observations,but the distinct ones are 23,133, so we will drop 73 observations

In [47]:
%%bigquery
CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.2021_unique`AS
SELECT *
FROM (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY NAME ORDER BY NAME) AS row_num
    FROM `ba-775-b03.BA_775_Dataset.2021` 
) AS subquery
WHERE row_num = 1;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

The table that has information of the payroll in 2021, has 22,546 observations,but the distinct ones are 22,472, so we will drop 74 observations

In [48]:
%%bigquery
CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.2020_unique`AS
SELECT *
FROM (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY NAME ORDER BY NAME) AS row_num
    FROM `ba-775-b03.BA_775_Dataset.2020` 
) AS subquery
WHERE row_num = 1;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

The table that has information of the payroll in 2020, has 21,858 observations,but the distinct ones are 21,785, so we will drop 73 observations

In [49]:
%%bigquery
CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.2019_unique`AS
SELECT *
FROM (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY NAME ORDER BY NAME) AS row_num
    FROM `ba-775-b03.BA_775_Dataset.2019` 
) AS subquery
WHERE row_num = 1;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

The table that has information of the payroll in 2019, has 23,312 observations,but the distinct ones are 23,232, so we will drop 80 observations

In [50]:
%%bigquery
CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.2018_unique`AS
SELECT *
FROM (
    SELECT *,
           ROW_NUMBER() OVER (PARTITION BY NAME ORDER BY NAME) AS row_num
    FROM `ba-775-b03.BA_775_Dataset.2018` 
) AS subquery
WHERE row_num = 1;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

The table that has information of the payroll in 2018, has 23,603 observations,but the distinct ones are 23,525, so we will drop 78 observations

In [51]:
%%bigquery 

CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.data_merged_raw` AS

(SELECT
  REPLACE(CAST(NAME AS STRING), ',', ' ') AS name,
  DEPARTMENT_NAME AS department_name,
  TITLE AS job_title,
  CAST(REGULAR AS FLOAT64) AS regular,
  CAST(RETRO AS FLOAT64) AS retro,
  CAST(OTHER AS FLOAT64) AS other,
  CAST(OVERTIME AS FLOAT64) AS overtime,
  CAST(INJURED AS FLOAT64) AS injured,
  CAST(DETAIL AS FLOAT64) AS detail,
  CAST(QUINN_EDUCATION AS FLOAT64) AS education,
  CAST(TOTAL__GROSS AS FLOAT64) AS total_earnings,
  CAST(POSTAL AS TINYINT) AS zip_code,
  2022 AS year
FROM
  `ba-775-b03.BA_775_Dataset.2022_unique`


UNION ALL


SELECT
  REPLACE(CAST(NAME AS STRING), ',', ' ') AS name,
  DEPARTMENT_NAME AS department_name,
  TITLE AS job_title,
  CAST(REGULAR AS FLOAT64) AS regular,
  CAST(RETRO AS FLOAT64) AS retro,
  CAST(OTHER AS FLOAT64) AS other,
  CAST(OVERTIME AS FLOAT64) AS overtime,
  CAST(INJURED AS FLOAT64) AS injured,
  CAST(DETAIL AS FLOAT64) AS detail,
  CAST(QUINN_EDUCATION_INCENTIVE AS FLOAT64) AS education,
  CAST(TOTAL_GROSS AS FLOAT64) AS total_earnings,
  CAST(POSTAL AS TINYINT) AS zip_code,
  2021 AS year
FROM
  `ba-775-b03.BA_775_Dataset.2021_unique`


UNION ALL


SELECT
  REPLACE(CAST(NAME AS STRING), ',', ' ') AS name,
  DEPARTMENT_NAME AS department_name,
  TITLE AS job_title,
  CAST(_REGULAR_ AS FLOAT64) AS regular,
  CAST(_RETRO_ AS FLOAT64) AS retro,
  CAST(_OTHER_ AS FLOAT64) AS other,
  CAST(_OVERTIME_ AS FLOAT64) AS overtime,
  CAST(_INJURED_ AS FLOAT64) AS injured,
  CAST(_DETAIL_ AS FLOAT64) AS detail,
  CAST(_QUINN___EDUCATION_INCENTIVE_ AS FLOAT64) AS education,
  CAST(_TOTAL_EARNINGS_ AS FLOAT64) AS total_earnings,
  CAST(POSTAL AS TINYINT) AS zip_code,
  2020 AS year
FROM
  `ba-775-b03.BA_775_Dataset.2020_unique`


UNION ALL


SELECT
  REPLACE(CAST(NAME AS STRING), ',', ' ') AS name,
  DEPARTMENT_NAME AS department_name,
  TITLE AS job_title,
  CAST(_REGULAR_ AS FLOAT64) AS regular,
  CAST(_RETRO_ AS FLOAT64) AS retro,
  CAST(_OTHER_ AS FLOAT64) AS other,
  CAST(_OVERTIME_ AS FLOAT64) AS overtime,
  CAST(_INJURED_ AS FLOAT64) AS injured,
  CAST(_DETAIL_ AS FLOAT64) AS detail,
  CAST(_QUINN_EDUCATION_INCENTIVE_ AS FLOAT64) AS education,
  CAST(TOTAL_EARNINGS AS FLOAT64) AS total_earnings,
  CAST(POSTAL AS TINYINT) AS zip_code,
  2019 AS year
FROM
  `ba-775-b03.BA_775_Dataset.2019_unique`


UNION ALL


SELECT
  REPLACE(CAST(NAME AS STRING), ',', ' ') AS name,
  DEPARTMENT_NAME AS department_name,
  TITLE AS job_title,
  CAST(REGULAR AS FLOAT64) AS regular,
  CAST(RETRO AS FLOAT64) AS retro,
  CAST(OTHER AS FLOAT64) AS other,
  CAST(OVERTIME AS FLOAT64) AS overtime,
  CAST(INJURED AS FLOAT64) AS injured,
  CAST(DETAIL AS FLOAT64) AS detail,
  CAST(QUINN_EDUCATION_INCENTIVE AS FLOAT64) AS education,
  CAST(TOTAL_EARNINGS AS FLOAT64) AS total_earnings,
  CAST(POSTAL AS TINYINT) AS zip_code,
  2018 AS year
FROM
  `ba-775-b03.BA_775_Dataset.2018_unique`);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

New Merged Table with data from 2018 - 2022

In [52]:
%%bigquery
SELECT*
FROM `ba-775-b03.BA_775_Dataset.data_merged_raw`

Query is running:   0%|          |

Downloading:   0%|          |

name           department_name  \
0                 Conlin Jane F        BPS Health Service   
1       Decius Fabiola Rachelle       Quincy Upper School   
2                Desances Kerly     BPS Madison Park High   
3                 Buendia Maria       Higginson/Lewis K-8   
4                   Mroz Joel P      BPS Ellis Elementary   
...                         ...                       ...   
114142         Danilecki John H  Boston Police Department   
114143         Bannish Kaitlynn     BPS Special Education   
114144         Bannish Kaitlynn     BPS Special Education   
114145         Bannish Kaitlynn     BPS Special Education   
114146         Santiago Natalia      BPS Adams Elementary   

                   job_title    regular    retro     other  overtime  injured  \
0                      Nurse  115017.59      NaN       NaN       NaN      NaN   
1                    Teacher   80064.12      NaN   6716.06       NaN      NaN   
2                    Teacher   88652.46      NaN    702.92       NaN      NaN   
3                    Teacher  128330.77      NaN   3612.22       NaN      NaN   
4                    Teacher   20405.78      NaN       NaN       NaN      NaN   
...                      ...        ...      ...       ...       ...      ...   
114142        Police Captain  157595.81      NaN  24504.13  49388.68  6504.89   
114143  ABA Specialist (BTU)   48156.31  1399.39       NaN       NaN      NaN   
114144  ABA Specialist (BTU)   55716.60      NaN       NaN       NaN      NaN   
114145  ABA Specialist (BTU)   49985.66   984.72    346.50       NaN      NaN   
114146               Teacher   55866.12      NaN    800.00       NaN      NaN   

         detail  education  total_earnings  zip_code  year  
0           NaN        NaN       115017.59      <NA>  2020  
1           NaN        NaN        86780.18      <NA>  2020  
2           NaN        NaN        89355.38      <NA>  2020  
3           NaN        NaN       131942.99      <NA>  2020  
4           NaN        NaN        20405.78      <NA>  2020  
...         ...        ...             ...       ...   ...  
114142  64800.0   41025.08       343818.59      2559  2021  
114143      NaN        NaN        49555.70      2815  2018  
114144      NaN        NaN        55716.60      2815  2020  
114145      NaN        NaN        51316.88      2815  2019  
114146      NaN        NaN        56666.12     33023  2020  

[114147 rows x 13 columns]

In [53]:
%%bigquery
SELECT
  COUNTIF(name IS NULL) AS name,
  COUNTIF(department_name IS NULL) AS department_name,
  COUNTIF(job_title IS NULL) AS job_title,
  COUNTIF(regular IS NULL) AS regular,
  COUNTIF(retro IS NULL) AS retro,
  COUNTIF(other IS NULL) AS other,
  COUNTIF(overtime IS NULL) AS overtime,
  COUNTIF(injured IS NULL) AS injured,
  COUNTIF(detail IS NULL) AS detail,
  COUNTIF(education IS NULL) AS education,
  COUNTIF(total_earnings IS NULL) AS total_earnings,
  COUNTIF(zip_code IS NULL) AS zip_code,
  COUNTIF(year IS NULL) AS year
FROM `ba-775-b03.BA_775_Dataset.data_merged_raw`

Query is running:   0%|          |

Downloading:   0%|          |

name  department_name  job_title  regular  retro  other  overtime  injured  \
0     0                0          0     4751  84277  43754     79977   107213   

   detail  education  total_earnings  zip_code  year  
0  103262     107329               0        93     0

Write more info about null values

In [54]:
%%bigquery
UPDATE `ba-775-b03.BA_775_Dataset.data_merged_raw`
SET
  regular = COALESCE(regular, 0),
  retro = COALESCE(retro, 0),
  other = COALESCE(other, 0),
  overtime = COALESCE(overtime, 0),
  injured = COALESCE(injured, 0),
  detail = COALESCE(detail, 0),
  education = COALESCE(education, 0),
  total_earnings = COALESCE(total_earnings, 0),
  zip_code = COALESCE(zip_code, 0)
WHERE TRUE;

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [55]:
%%bigquery
SELECT
  COUNTIF(name IS NULL) AS name,
  COUNTIF(department_name IS NULL) AS department_name,
  COUNTIF(job_title IS NULL) AS job_title,
  COUNTIF(regular IS NULL) AS regular,
  COUNTIF(retro IS NULL) AS retro,
  COUNTIF(other IS NULL) AS other,
  COUNTIF(overtime IS NULL) AS overtime,
  COUNTIF(injured IS NULL) AS injured,
  COUNTIF(detail IS NULL) AS detail,
  COUNTIF(education IS NULL) AS education,
  COUNTIF(total_earnings IS NULL) AS total_earnings,
  COUNTIF(zip_code IS NULL) AS zip_code,
  COUNTIF(year IS NULL) AS year
FROM `ba-775-b03.BA_775_Dataset.data_merged_raw`

Query is running:   0%|          |

Downloading:   0%|          |

name  department_name  job_title  regular  retro  other  overtime  injured  \
0     0                0          0        0      0      0         0        0   

   detail  education  total_earnings  zip_code  year  
0       0          0               0         0     0

Replacing all the null values

In [56]:
%%bigquery

CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.data_merged_cleaned_joined_zip` AS
(
SELECT *
FROM `ba-775-b03.BA_775_Dataset.data_merged_raw` AS earnings
LEFT JOIN `ba-775-b03.BA_775_Dataset.MA_zipcodeda` as zip_code
ON earnings.zip_code = zip_code.zip);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [57]:
%%bigquery
SELECT *
FROM `ba-775-b03.BA_775_Dataset.data_merged_cleaned_joined_zip`

Query is running:   0%|          |

Downloading:   0%|          |

name               department_name  \
0                     Gates Joshua       BPS Human Resource Team   
1                 Zawislak Wanda I            BPS Timilty Middle   
2                     Plummer Abby                    BPS Equity   
3       Pattz Desmond Trevor Wayne      BPS Higginson Elementary   
4                Ambroise Michel c  BPS Substitute Teachers/Nurs   
...                            ...                           ...   
114142          Carrasquillo Mayra          Food & Nutrition Svc   
114143         Serra Christopher A        Boston Fire Department   
114144              Sullivan Carla      Boston Police Department   
114145                 Tavares Kim      Boston Police Department   
114146            Danilecki John H      Boston Police Department   

                   job_title    regular    retro     other  overtime  injured  \
0                    Teacher   89306.58  2311.50   1000.00      0.00     0.00   
1                    Teacher  106236.26  2872.00      0.00      0.00     0.00   
2                   Intern 3    4112.50     0.00      0.00      0.00     0.00   
3           Paraprofessional       0.00   665.03   3077.68      0.00     0.00   
4         Substitute Teacher   15661.65    88.51      0.00      0.00     0.00   
...                      ...        ...      ...       ...       ...      ...   
114142  Cafeteria Manager 12   46842.35     0.00   3100.00     67.42     0.00   
114143          Fire Fighter   22775.47     0.00  65516.36   2095.21     0.00   
114144      Police Detective  106191.97     0.00  12405.70  15372.25  1160.57   
114145        Police Officer  105343.12     0.00   3450.85  58362.74     0.00   
114146        Police Captain  164100.70     0.00  24504.13  71315.48     0.00   

         detail  education  ...   zip      type  primary_city  state  \
0           0.0       0.00  ...  <NA>      None          None   None   
1           0.0       0.00  ...  <NA>      None          None   None   
2           0.0       0.00  ...  <NA>      None          None   None   
3           0.0       0.00  ...  <NA>      None          None   None   
4           0.0       0.00  ...  <NA>      None          None   None   
...         ...        ...  ...   ...       ...           ...    ...   
114142      0.0       0.00  ...  2302  STANDARD      Brockton     MA   
114143      0.0       0.00  ...  2047    PO BOX      Humarock     MA   
114144  55200.0   26838.10  ...  2047    PO BOX      Humarock     MA   
114145      0.0       0.00  ...  2303    PO BOX      Brockton     MA   
114146  61200.0   41025.08  ...  2559  STANDARD      Pocasset     MA   

                   county          timezone world_region country latitude  \
0                    None              None         None    None      NaN   
1                    None              None         None    None      NaN   
2                    None              None         None    None      NaN   
3                    None              None         None    None      NaN   
4                    None              None         None    None      NaN   
...                   ...               ...          ...     ...      ...   
114142    Plymouth County  America/New_York           NA      US    42.09   
114143    Plymouth County  America/New_York           NA      US    42.15   
114144    Plymouth County  America/New_York           NA      US    42.15   
114145    Plymouth County  America/New_York           NA      US    42.08   
114146  Barnstable County  America/New_York           NA      US    41.69   

       longitude  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
114142    -71.00  
114143    -70.70  
114144    -70.70  
114145    -71.02  
114146    -70.63  

[114147 rows x 23 columns]

In [58]:
%%bigquery

SELECT
  COUNTIF(name IS NULL) AS name,
  COUNTIF(department_name IS NULL) AS department_name,
  COUNTIF(job_title IS NULL) AS job_title,
  COUNTIF(regular IS NULL) AS regular,
  COUNTIF(retro IS NULL) AS retro,
  COUNTIF(other IS NULL) AS other,
  COUNTIF(overtime IS NULL) AS overtime,
  COUNTIF(injured IS NULL) AS injured,
  COUNTIF(detail IS NULL) AS detail,
  COUNTIF(education IS NULL) AS education,
  COUNTIF(total_earnings IS NULL) AS total_earnings,
  COUNTIF(zip_code IS NULL) AS zip_code,
  COUNTIF(year IS NULL) AS year,
  COUNTIF(zip IS NULL) AS zip
FROM `ba-775-b03.BA_775_Dataset.data_merged_cleaned_joined_zip`;

Query is running:   0%|          |

Downloading:   0%|          |

name  department_name  job_title  regular  retro  other  overtime  injured  \
0     0                0          0        0      0      0         0        0   

   detail  education  total_earnings  zip_code  year   zip  
0       0          0               0         0     0  1730

1731 outside mass; add more insights

In [59]:
%%bigquery

DELETE FROM `ba-775-b03.BA_775_Dataset.data_merged_cleaned_joined_zip`
WHERE ZIP IS NULL;

CREATE OR REPLACE TABLE `ba-775-b03.BA_775_Dataset.data_final` AS
(SELECT *
FROM `ba-775-b03.BA_775_Dataset.data_merged_cleaned_joined_zip`);

Query is running:   0%|          |

Empty DataFrame
Columns: []
Index: []

In [60]:
%%bigquery

SELECT *
FROM `ba-775-b03.BA_775_Dataset.data_final`

Query is running:   0%|          |

Downloading:   0%|          |

name                department_name  \
0            Flores Ty-Juana L       Office of Arts & Culture   
1            Flores Ty-Juana L       Office of Arts & Culture   
2            Smith Wesley Drew     Treasury-Treasury Division   
3            Smith Wesley Drew     Treasury-Treasury Division   
4               Costello Megan            Women's Advancement   
...                        ...                            ...   
112412  Conward-Lewis Breannah  Youth Engagement & Employment   
112413  Conward-Lewis Breannah  Youth Engagement & Employment   
112414    Raciukaitis Lina Ona          Boston Public Library   
112415    Raciukaitis Lina Ona          Boston Public Library   
112416    Raciukaitis Lina Ona          Boston Public Library   

                             job_title    regular   retro     other  overtime  \
0                                  Dir    5006.83    0.00      0.00      0.00   
1                                  Dir   30040.96    0.00      0.00      0.00   
2                                  Dir   86307.70    0.00  19450.55      0.00   
3                                  Dir  130576.93    0.00   2500.00      0.00   
4                                  Dir   53540.36  880.96   4756.04      0.00   
...                                ...        ...     ...       ...       ...   
112412  Yth & Career Development Coord    2799.91   98.13   3919.89     53.33   
112413  Yth & Career Development Coord   37292.24    0.00      0.00   1239.96   
112414  Yth Educational Outreach Coord   35734.54    0.00      0.00      0.00   
112415  Yth Educational Outreach Coord   18726.98    0.00   5268.71      0.00   
112416  Yth Educational Outreach Coord   30715.80    0.00      0.00      0.00   

        injured  detail  education  ...   zip      type  primary_city  state  \
0           0.0     0.0        0.0  ...  2121  STANDARD    Dorchester     MA   
1           0.0     0.0        0.0  ...  2121  STANDARD    Dorchester     MA   
2           0.0     0.0        0.0  ...  2122  STANDARD    Dorchester     MA   
3           0.0     0.0        0.0  ...  2122  STANDARD    Dorchester     MA   
4           0.0     0.0        0.0  ...  2122  STANDARD    Dorchester     MA   
...         ...     ...        ...  ...   ...       ...           ...    ...   
112412      0.0     0.0        0.0  ...  2125  STANDARD    Dorchester     MA   
112413      0.0     0.0        0.0  ...  2125  STANDARD    Dorchester     MA   
112414      0.0     0.0        0.0  ...  2134  STANDARD       Allston     MA   
112415      0.0     0.0        0.0  ...  2134  STANDARD       Allston     MA   
112416      0.0     0.0        0.0  ...  2134  STANDARD       Allston     MA   

                county          timezone world_region country latitude  \
0       Suffolk County  America/New_York           NA      US    42.31   
1       Suffolk County  America/New_York           NA      US    42.31   
2       Suffolk County  America/New_York           NA      US    42.29   
3       Suffolk County  America/New_York           NA      US    42.29   
4       Suffolk County  America/New_York           NA      US    42.29   
...                ...               ...          ...     ...      ...   
112412  Suffolk County  America/New_York           NA      US    42.32   
112413  Suffolk County  America/New_York           NA      US    42.32   
112414  Suffolk County  America/New_York           NA      US    42.35   
112415  Suffolk County  America/New_York           NA      US    42.35   
112416  Suffolk County  America/New_York           NA      US    42.35   

       longitude  
0         -71.09  
1         -71.09  
2         -71.04  
3         -71.04  
4         -71.04  
...          ...  
112412    -71.06  
112413    -71.06  
112414    -71.13  
112415    -71.13  
112416    -71.13  

[112417 rows x 23 columns]

By dropping all the null values, we have eliminated 1730 observations, the ones that do not registered a valid zip code

# **Questions and Analysis**

In this section, we will analyze the cleaned dataset to gain insights.

## Storyline:

## **Section 1: Year-wise breakdown**

### **Question 1**

### **Question 2**

### **Question 3**

### **Question 4**

## **Section 2: Department-wise breakdown**

### **Question 1:** How do the number of employees and median salaries across different departments correlate with the allocation of their pay components (regular, retroactive, other, and overtime) in millions, and which department exhibits the highest median earnings per employee, indicating potentially efficient budget utilization?

In [61]:
%%bigquery
WITH MedianSalaries AS (
SELECT
  department_name,
  APPROX_QUANTILES((regular + IFNULL(retro, 0) + IFNULL(other, 0) + IFNULL(overtime, 0)), 2)[OFFSET(1)] AS median_salary
FROM `ba-775-b03.BA_775_Dataset.data_final`
GROUP BY
  department_name
)
SELECT
ms.department_name,
COUNT(e.name) AS number_of_employees,
SUM(e.regular) / 1000000 AS total_regular_pay_in_millions,
SUM(e.retro) / 1000000 AS total_retro_pay_in_millions,
SUM(e.other) / 1000000 AS total_other_pay_in_millions,
SUM(e.overtime) / 1000000 AS total_overtime_pay_in_millions,
ROUND(SUM(e.regular + IFNULL(e.retro, 0) + IFNULL(e.other, 0) + IFNULL(e.overtime, 0)) / 1000000, 2) AS total_earnings_in_millions,
ROUND(ms.median_salary / 1000000, 2) AS median_salary_in_millions
FROM `ba-775-b03.BA_775_Dataset.data_final` e
JOIN
MedianSalaries ms ON e.department_name = ms.department_name
GROUP BY
ms.department_name, ms.median_salary
ORDER BY
total_earnings_in_millions DESC
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

department_name  number_of_employees  \
0        Boston Police Department                15514   
1          Boston Fire Department                 8540   
2           BPS Special Education                 4097   
3         BPS Facility Management                 2948   
4           Boston Public Library                 2496   
5         Public Works Department                 2057   
6  Boston Cntr - Youth & Families                 2265   
7                Traffic Division                 1761   
8           BPS Madison Park High                 1143   
9      Inspectional Services Dept                 1113   

   total_regular_pay_in_millions  total_retro_pay_in_millions  \
0                    1184.136641                     9.113784   
1                     946.028466                    17.779547   
2                     256.370650                     2.716519   
3                     132.395731                     1.255610   
4                     137.077742                     1.834487   
5                     105.470333                     1.102356   
6                     105.058821                     0.817317   
7                      95.965151                     0.895647   
8                      86.637333                     0.768595   
9                      76.165305                     0.788774   

   total_other_pay_in_millions  total_overtime_pay_in_millions  \
0                    85.359261                      380.489221   
1                    41.359242                      169.558076   
2                     5.654851                        0.104827   
3                     5.361500                       10.443515   
4                     3.032512                        2.641883   
5                     2.369946                       21.982451   
6                     2.320374                        1.516634   
7                     2.047477                        8.156571   
8                     3.074336                        0.056003   
9                     2.788521                        8.716531   

   total_earnings_in_millions  median_salary_in_millions  
0                     1659.10                       0.11  
1                     1174.73                       0.14  
2                      264.85                       0.05  
3                      149.46                       0.06  
4                      144.59                       0.06  
5                      130.93                       0.06  
6                      109.71                       0.05  
7                      107.06                       0.06  
8                       90.54                       0.09  
9                       88.46                       0.08

**Answer**: 
A higher median salary can indicate a department's commitment to competitive compensation, which can be essential for retaining talent. For example, the Boston Fire Department shows a median salary of USD 0.14 million, which is higher than that of the Boston Police Department at USD 0.11 million, despite the latter having a higher total earnings figure. This might suggest that the Fire Department has a more concentrated distribution of compensation towards the median, which could be interpreted as a more equitable pay structure.

### **Question 2:** In the analysis of public sector salaries within the Boston area, what patterns emerge from the top earnings across different departments, and how do regular, retroactive, other compensations, and overtime contribute to the total earnings of the highest-paid employees? Specifically, how do the earnings of the top officials in the Boston Police Department compare to those in the Boston Fire Department, and what might this indicate about the compensation structures and fiscal management within these public institutions?

In [62]:
%%bigquery
SELECT
name,
department_name,
job_title,
regular,
retro,
other,
overtime,
ROUND((regular + IFNULL(retro, 0) + IFNULL(other, 0) + IFNULL(overtime, 0)), 2) AS total_earnings
FROM `ba-775-b03.BA_775_Dataset.data_final`
ORDER BY
total_earnings DESC
LIMIT 10;

Query is running:   0%|          |

Downloading:   0%|          |

name           department_name                job_title  \
0        Beckers Richard  Boston Police Department           Police Officer   
1  McGowan Jacqueline M.  Boston Police Department           Police Officer   
2         Gavin Donna M.  Boston Police Department  Police Lieutenant (Det)   
3         Harris Shawn N  Boston Police Department           Police Officer   
4      Williams David C.  Boston Police Department           Police Officer   
5         Jones Ronnie C  Boston Police Department           Police Officer   
6       Downing George C  Boston Police Department           Police Officer   
7      Washington Walter  Boston Police Department           Police Officer   
8           Dempsey John    Boston Fire Department       Commissioner (BFD)   
9      Washington Walter  Boston Police Department           Police Officer   

     regular  retro       other  overtime  total_earnings  
0       0.00    0.0  1264843.63      0.00      1264843.63  
1       0.00    0.0  1252990.81      0.00      1252990.81  
2   53226.90    0.0  1045845.44    463.52      1099535.86  
3  517789.82    0.0    83369.34  48199.74       649358.90  
4   95184.68    0.0   503025.78  45176.88       643387.34  
5  462149.98    0.0   114973.46      0.00       577123.44  
6  417664.03    0.0    78599.22  29567.40       525830.65  
7  313154.27    0.0    99893.17  45337.30       458384.74  
8  133653.78    0.0   312752.53      0.00       446406.31  
9  100963.38    0.0   211900.28  67849.66       380713.32

**Answer**:
Top Earnings Overview: Richard Beckers and Jacqueline M. McGowan of the Boston Police Department’s total earnings are significantly higher than their colleagues due to large 'other' payments, with both having no regular or overtime pay listed. This raises questions about the nature of these 'other' payments and whether they reflect bonus compensations, back pay, or other non-standard financial arrangements.

Salary Discrepancies: Donna M. Gavin, a Police Lieutenant, shows a balanced composition of pay with a modest regular pay and minimal overtime, indicating a structured compensation with less reliance on variable pay. In contrast, Shawn N. Harris, a Police Officer, has an exceptionally high regular pay and a large sum of overtime pay, highlighting disparities that may arise from different roles, special duties, or administrative errors that warrant further review.

Overtime: Overtime pay, seen in significant amounts for Harris and Williams, suggests these individuals are either extensively involved in operations requiring long hours or there may be systematic overtime budgeting issues within the department. The large variance in overtime pay among the top earners necessitates a deeper look into departmental staffing and workload distribution.

Comparative Analysis: The Boston Fire Department's top earner, John Dempsey, shows no overtime pay, which starkly contrasts with the police department's pattern. This could reflect differing operational demands or more stringent overtime controls within the fire department.

### **Question 3:** What does the salary structure reveal about the distribution of earnings among the highest-paid individuals in various departments of the Boston public sector, and how do these figures compare within and across these departments?

In [63]:
%%bigquery
WITH RankedEarnings AS (
 SELECT
   name,
    year, # I have added year, i dont know if you think its correct to add it 
   department_name,
   job_title,
   ROUND(regular + IFNULL(retro, 0) + IFNULL(other, 0) + IFNULL(overtime, 0), 2) AS total_earnings,
   DENSE_RANK() OVER (PARTITION BY department_name ORDER BY ROUND(regular + IFNULL(retro, 0) + IFNULL(other, 0) + IFNULL(overtime, 0), 2) DESC) AS rank
FROM `ba-775-b03.BA_775_Dataset.data_final`
)
SELECT
 name,
 year,
 department_name,
 job_title,
 total_earnings
FROM
 RankedEarnings
WHERE
 rank = 1;

Query is running:   0%|          |

Downloading:   0%|          |

name  year               department_name  \
0           Cahill Karen M  2022        BPS Everett Elementary   
1             Clark Anne R  2022       BPS Boston Arts Academy   
2           Greenwood Lori  2022           Enrollment Services   
3       Holmes Christine M  2018  BPS Another Course To Colleg   
4    Cauley Charles Eudene  2021           BPS Dearborn Middle   
..                     ...   ...                           ...   
254   Cirasuolo Stephen A.  2022           Quincy Upper School   
255        Smith Jomila C.  2022        Research Assess & Eval   
256     Henderson Walter K  2018     Mattapan Early Elementary   
257   McMahon Patricia Ann  2022             Registry Division   
258        Santiago Robert  2022            Veterans' Services   

                job_title  total_earnings  
0    Principal Elementary       170759.83  
1          Head of School       161278.43  
2    Director (NonAcad) D       119523.06  
3                 Teacher       153613.20  
4     Instructional Coach       161686.94  
..                    ...             ...  
254        Head of School       155392.33  
255  Principal Elementary       143812.84  
256  Principal Elementary       136046.83  
257        City Registrar       119346.22  
258    Commissioner (Vet)       110057.72  

[259 rows x 5 columns]

**Answer**: Diverse Compensation Structures: The dataset showcases a range of total earnings, from Imani Monet Penn's USD 4,640 as an intern in the BPS Development to Leslie L Gant's USD 193,233.08 as a Principal at BPS Tynan Elementary. This variance indicates a broad spectrum of roles and the associated value placed on these positions within the public sector.

Top Earners by Department: High-ranking roles, such as Jill E Carter as Assistant Superintendent and Kristen McCosh as Commissioner, command substantial earnings, at USD 168,470.83 and USD 119,961.5, respectively. These figures reflect the significant fiscal responsibility and leadership expectations associated with their positions.

Compensation and Job Titles: There's a notable correlation between job titles and earnings. Leadership positions such as elementary principals and commissioners are among the top earners, which aligns with the additional responsibilities and expertise required for these roles.

Comparative Earnings Analysis: Comparing departments, we see a Commissioner in the Commission for Persons with Disabilities earning USD 119,961.5, whereas a Commissioner of Elderly Affairs earns USD 121,577.14, suggesting parity in pay for commissioners across different public service domains.

Anomalies and Outliers: The standout anomaly is the earnings for the 'Retirees' category, with Glenn L Brewington receiving only USD 3,020.72, a figure that dramatically deviates from the rest and could represent part-time work, a pension payout, or an administrative error.

Inconsistencies in Pay: Marlene M Gibbs, a Teacher, earns USD 158,371.55, which is quite high relative to the common pay range for teachers. This could be due to additional duties, long tenure, or special qualifications warranting further examination.

Fiscal Oversight Implications: The data indicates a need for fiscal oversight to ensure that salaries are not only competitive but also justified based on the roles and responsibilities within each department.


## **Section 3: Income components breakdown**

### **Question 1:** Find the top 10 employees who contributed the most to the regular earnings component and which department are they in?

In [64]:
%%bigquery
WITH TopRegularContributors AS (
 SELECT
   name,
   department_name,
   regular,
   RANK() OVER (ORDER BY regular DESC) AS ranking
FROM 
    `ba-775-b03.BA_775_Dataset.data_final`
)
SELECT
 name,
 department_name,
 regular
FROM
 TopRegularContributors
WHERE
 ranking <= 10
ORDER BY regular DESC;

Query is running:   0%|          |

Downloading:   0%|          |

name           department_name    regular
0     Harris Shawn N  Boston Police Department  517789.82
1     Jones Ronnie C  Boston Police Department  462149.98
2   Downing George C  Boston Police Department  417664.03
3  Washington Walter  Boston Police Department  313154.27
4     Long Gregory P  Boston Police Department  290703.31
5  Cassellius Brenda            Superintendent  279999.98
6  Cassellius Brenda            Superintendent  279999.98
7       Dempsey John    Boston Fire Department  262657.41
8   McMahon Joseph G    Boston Fire Department  262641.69
9   Walsh David John    Boston Fire Department  262134.86

**Answer**: Out of the top 10 employees , 50% of the employees are from the Police department whose regular components are the highest, 30% from the top 10 belong to the fire department and the rest 20% are Superintendents.

### **Question 2:** Which department had paid the most to their employees for each specific year?

In [65]:
%%bigquery
WITH DepartmentTotalEarnings AS (
 SELECT
   department_name,
   year,
   SUM(total_earnings) AS total_earnings_sum
FROM 
    `ba-775-b03.BA_775_Dataset.data_final`
 WHERE
   year IN (2018, 2019, 2020, 2021,2022)
GROUP BY
   department_name,
year
)
SELECT
 d.department_name,
 d.year,
 d.total_earnings_sum
FROM
 DepartmentTotalEarnings d
JOIN ( SELECT
year,
   MAX(total_earnings_sum) AS max_earnings
 FROM
   DepartmentTotalEarnings
GROUP BY
   year
) max_earnings_per_year
ON d.year = max_earnings_per_year.year
AND d.total_earnings_sum = max_earnings_per_year.max_earnings;

Query is running:   0%|          |

Downloading:   0%|          |

department_name  year  total_earnings_sum
0  Boston Police Department  2021        4.040525e+08
1  Boston Police Department  2020        4.112909e+08
2  Boston Police Department  2022        4.008621e+08
3  Boston Police Department  2018        4.125141e+08
4  Boston Police Department  2019        4.105958e+08

**Answer**: The Boston Police Department consistently paid the highest total earnings to its employees across the years 2018 to 2022. The department maintained its position as the top payer each year, also we notice a declining trend in total earnings observed over this period.

### **Question 3:** Which were the top 3 job title that got the highest total earnings in the past 5 years?

In [66]:
%%bigquery
WITH JobTitleTotalEarnings AS (
 SELECT
   job_title,
   department_name,
   SUM(total_earnings) AS total_earnings_sum
FROM 
    `ba-775-b03.BA_775_Dataset.data_final`
 WHERE
   year BETWEEN EXTRACT(YEAR FROM CURRENT_DATE()) - 4 AND EXTRACT(YEAR FROM
CURRENT_DATE())
 GROUP BY
   job_title, department_name
)
SELECT
 job_title,
 department_name,
 FORMAT('%t', total_earnings_sum / 1000000) AS total_earnings_in_millions
FROM
 JobTitleTotalEarnings
ORDER BY
 total_earnings_sum DESC
LIMIT 3;

Query is running:   0%|          |

Downloading:   0%|          |

job_title           department_name total_earnings_in_millions
0    Police Officer  Boston Police Department         696.67158547000349
1      Fire Fighter    Boston Fire Department         464.88350213999996
2  Police Detective  Boston Police Department         202.94132968000031

**Answer**: The top three job titles with the highest total earnings in millions are Police Officer and Police Detective from the Boston Police Department, along with Fire Fighter from the Boston Fire Department. This suggests that law enforcement roles, particularly within the police department, tend to have higher overall earnings compared to other positions, with Police Officer being the highest.

### **Question 4:** Which income component has the most impact after regular on an employee’s gross earnings for the top 3 departments with highest pay out.

In [67]:
%%bigquery
SELECT
 department_name,
 SUM(regular) / SUM(total_earnings) * 100 AS regular_percentage,
 SUM(retro) / SUM(total_earnings) * 100 AS retro_percentage,
 SUM(other) / SUM(total_earnings) * 100 AS other_percentage,
 SUM(overtime) / SUM(total_earnings) * 100 AS overtime_percentage,
 SUM(injured) / SUM(total_earnings) * 100 AS injured_percentage,
 SUM(detail) / SUM(total_earnings) * 100 AS detail_percentage,
 SUM(education) / SUM(total_earnings) * 100 AS education_percentage,
FROM 
    `ba-775-b03.BA_775_Dataset.data_final`
GROUP BY
 department_name
ORDER BY
 SUM(total_earnings) DESC
LIMIT 3;

Query is running:   0%|          |

Downloading:   0%|          |

department_name  regular_percentage  retro_percentage  \
0  Boston Police Department           58.065400          0.446904   
1    Boston Fire Department           73.410363          1.379666   
2     BPS Special Education           96.748382          1.025152   

   other_percentage  overtime_percentage  injured_percentage  \
0          4.185682            18.657694            5.227077   
1          3.209414            13.157447            5.603053   
2          2.134010             0.039559            0.052897   

   detail_percentage  education_percentage  
0           6.636647              6.781138  
1           3.238120              0.002564  
2           0.000000              0.000000

**Answer**: 
1. **Police Department**: The highest variable pay comes from Overtime work that the people in the Police department do.
2. **Fire Department**: Similar to the police department the highest variable for the fire department comes from Overtime pay that employees do.

* Employees in the Police and Fire department are driven by the potential for increased earnings through performance-related bonuses,

3. **Boston Public Schools**: We don’t see any significant contribution from variable pay in the Public Schools department which implies that the individual’s income is not dependent on performance or external factors. There is job security and comparatively stable to others

## **Section 4: COVID-19’s Impact analysis**

In accordance with what is indicated in section 1 (Year-wise breakdown), it has been validated that in 2020 the lowest levels of hired workers were recorded. In this section we will analyze the previous (2018) and subsequent (2022) impact of COVID19 on workers in the government sector of Boston, for this, we will do an analysis at the Departmental and County level.

### **Question 1:** Which department experienced the most significant workforce reductions between 2018 and 2020, before the emergence of the COVID-19 pandemic?

In [68]:
%%bigquery
WITH sq1 AS (
    SELECT *
    FROM `ba-775-b03.BA_775_Dataset.data_final`
    WHERE year IN (2018, 2020)
)

, sq2 AS (
    SELECT year, department_name, county, COUNT(*) AS total_workers
    FROM sq1
    GROUP BY year, department_name, county
)

, sq3 AS (
    SELECT
        department_name,
        MAX(CASE WHEN year = 2018 THEN total_workers END) AS total_workers_2018,
        MAX(CASE WHEN year = 2020 THEN total_workers END) AS total_workers_2020
    FROM sq2
    GROUP BY department_name
)
SELECT
    RANK() OVER (ORDER BY total_workers_2020 - total_workers_2018) AS rank_var_2020_2018,
    department_name,
    total_workers_2020 - total_workers_2018 AS w_var_2020_2018
FROM sq3
WHERE total_workers_2020 IS NOT NULL AND total_workers_2018 IS NOT NULL
ORDER BY rank_var_2020_2018;

Query is running:   0%|          |

Downloading:   0%|          |

rank_var_2020_2018                 department_name  w_var_2020_2018
0                     1    BPS Substitute Teachers/Nurs             -273
1                     2  Boston Cntr - Youth & Families             -178
2                     3         BPS Facility Management             -102
3                     4                Parks Department              -55
4                     5    Workers Compensation Service              -43
..                  ...                             ...              ...
213                 214              BPS Lee Elementary               10
214                 214           Neighborhood Services               10
215                 216                       Perry K-8               11
216                 217             Teaching & Learning               17
217                 218        Boston Police Department               27

[218 rows x 3 columns]

**Answer**: The analysis of workforce changes between 2018 and 2020 reveals that the department most profoundly affected by reductions was BPS Substitute Teachers/Nursing, witnessing a substantial decline of 273 workers. Following closely behind are Boston Center for Youth & Families (-178), BPS Facility Management (-102), Parks Department (-55), and Workers Compensation Service (-43). Notably, regarding the findings in the second section, it is intriguing to observe that BPS Substitute Teachers/Nursing did not rank among the top ten departments with the highest workforce numbers.

### **Question 2:** Which department witnessed the most substantial increase in workforce hiring between 2020 and 2022, following the onset of the COVID-19 pandemic?

In [69]:
%%bigquery
WITH sq1 AS (
    SELECT *
    FROM `ba-775-b03.BA_775_Dataset.data_final`
    WHERE year IN (2020, 2022)
)

, sq2 AS (
    SELECT year, department_name, county, COUNT(*) AS total_workers
    FROM sq1
    GROUP BY year, department_name, county
)

, sq3 AS (
    SELECT
        department_name,
        MAX(CASE WHEN year = 2020 THEN total_workers END) AS total_workers_2020,
        MAX(CASE WHEN year = 2022 THEN total_workers END) AS total_workers_2022
    FROM sq2
    GROUP BY department_name
)

SELECT
    RANK() OVER (ORDER BY total_workers_2022 - total_workers_2020 DESC) AS rank_var_2022_2020,
    department_name,
    total_workers_2022 - total_workers_2020 AS w_var_2022_2020
FROM sq3
WHERE total_workers_2022 IS NOT NULL AND total_workers_2020  IS NOT NULL
ORDER BY rank_var_2022_2020;

Query is running:   0%|          |

Downloading:   0%|          |

rank_var_2022_2020           department_name  w_var_2022_2020
0                     1      Food & Nutrition Svc              428
1                     2    Boston Fire Department               87
2                     3     BPS Special Education               42
3                     4             BPS Athletics               34
4                     5      Health & Human Servs               25
..                  ...                       ...              ...
210                 211   BPS Human Resource Team              -29
211                 212          Mission Hill K-8              -33
212                 213          Jackson/Mann K-8              -38
213                 214  Boston Police Department              -39
214                 215          Traffic Division              -41

[215 rows x 3 columns]

**Answer**: 
The department that experienced the most significant increase in workforce hiring between 2020 and 2022, in the aftermath of the COVID-19 pandemic, is Food & Nutrition Services, marking an impressive surge with 428 new hires. Following closely behind are Boston Fire Department (87), BPS Special Education (42), BPS Athletics (34), and Health & Human Services (25). Although the Department of Food & Nutrition Services
does not appear in the top 10, these results highlight a notable expansion of staffing during the specified period, reflecting the organizational response to changing needs in the post-pandemic landscape.

### **Question 3:** Which county experienced the most significant worker layoffs between 2018 and 2020, prior to the onset of the COVID-19 pandemic?

In [70]:
%%bigquery
WITH sq1 AS (
    SELECT *
    FROM `ba-775-b03.BA_775_Dataset.data_final`
    WHERE year IN (2018, 2020)
)

, sq2 AS (
    SELECT year, county, COUNT(*) AS total_workers
    FROM sq1
    GROUP BY year, county
)

, sq3 AS (
    SELECT
        county,
        MAX(CASE WHEN year = 2018 THEN total_workers END) AS total_workers_2018,
        MAX(CASE WHEN year = 2020 THEN total_workers END) AS total_workers_2020
    FROM sq2
    GROUP BY county
)

, sq4 AS (
    SELECT
        county,
        total_workers_2020 - total_workers_2018 AS change_2020_2018
    FROM sq3
    WHERE total_workers_2020 IS NOT NULL AND total_workers_2018 IS NOT NULL
)

SELECT
    RANK() OVER (ORDER BY change_2020_2018) AS rank_change_2020_2018,
    county,
    change_2020_2018
FROM sq4
ORDER BY change_2020_2018;

Query is running:   0%|          |

Downloading:   0%|          |

rank_change_2020_2018             county  change_2020_2018
0                       1     Suffolk County             -1312
1                       2   Middlesex County              -248
2                       3     Norfolk County              -131
3                       4    Plymouth County               -30
4                       5  Barnstable County                -8
5                       6       Essex County                -7
6                       7     Hampden County                -4
7                       8       Dukes County                -2
8                       9   Berkshire County                -1
9                       9   Worcester County                -1
10                     11   Nantucket County                 0
11                     12     Bristol County                10

**Answer**: Suffolk County, which according to Section 1 results is the county with the highest percentage of employees, emerged as the county that faced the most substantial worker layoffs between 2018 and 2020, with a notable decrease of 1,312 workers. Middlesex County followed closely with a decrease of 248 workers, and Norfolk County saw a reduction of 131 workers. Additionally, Plymouth County experienced a decrease of 36 workers during this period. 

### **Question 4:** Which county underwent the most significant worker hiring between 2020 and 2022, following the COVID-19 pandemic?

In [71]:
%%bigquery
WITH sq1 AS (
    SELECT *
    FROM `ba-775-b03.BA_775_Dataset.data_final`
    WHERE year IN (2020, 2022)
)

, sq2 AS (
    SELECT year, county, COUNT(*) AS total_workers
    FROM sq1
    GROUP BY year, county
)

, sq3 AS (
    SELECT
        county,
        MAX(CASE WHEN year = 2020 THEN total_workers END) AS total_workers_2020,
        MAX(CASE WHEN year = 2022 THEN total_workers END) AS total_workers_2022
    FROM sq2
    GROUP BY county
)

, sq4 AS (
    SELECT
        county,
        total_workers_2022 - total_workers_2020 AS change_2022_2020
    FROM sq3
    WHERE total_workers_2022 IS NOT NULL AND total_workers_2020 IS NOT NULL
)

SELECT
    RANK() OVER (ORDER BY change_2022_2020 DESC) AS rank_change_2022_2020,
    county,
    change_2022_2020
FROM sq4
ORDER BY change_2022_2020 DESC;

Query is running:   0%|          |

Downloading:   0%|          |

rank_change_2022_2020             county  change_2022_2020
0                      1     Suffolk County               908
1                      2     Norfolk County               173
2                      3   Middlesex County               130
3                      4    Plymouth County                82
4                      5     Bristol County                43
5                      6       Essex County                37
6                      7   Worcester County                22
7                      8  Barnstable County                 4
8                      9   Berkshire County                 3
9                      9     Hampden County                 3

**Answer**: Suffolk County stands out as the county with the most significant worker hiring between 2020 and 2022, post-COVID-19 pandemic, witnessing an impressive increase of 908 workers. This surge in hiring is particularly noteworthy given Suffolk County's initial status as the county with the highest percentage of employees, as indicated in Section 1 results. Norfolk County follows with a notable hiring increase of 173 workers,and Middlesex with an increase of 130 workers.

# **Conclusion**

*Unraveling the US Economy (2000-2020)*

Our journey through the US economy from 2000 to 2020 unveiled some fascinating stories. One of the most striking tales was the dip during the 2008 recession, impacting both jobs and the overall economy. Although GDP and the number of businesses grew since 2000, job creation didn't keep up.

When we looked closer at the recession years, we discovered that some job sectors remained stable, like Agriculture, Utilities, and Education, while others, such as Administrative, Manufacturing, and Construction, faced more ups and downs.

As we traveled through states and their GDP, some states consistently stood out as leaders, like New York, California, and Texas. Florida, in particular, impressed us with its nearly doubled GDP over the course.

Lastly, we explored how firm closures and GDP are connected. It turns out that when businesses struggled during the recession, it had a noticeable impact on the economy. The more firms closed, the lower the GDP.

In simple terms, our analysis illuminated the US economy's ups and downs, showcasing its resilience, differences between job sectors, and the tightrope walk between business health and economic prosperity. These insights provide us with valuable guidance as we navigate the complex world of economics.


# **Acknowledgments**

We would like to acknowledge the following  

1. ChatGPT from OpenAI for its assistance during the course of this project. ChatGPT aided us in three key instances:
* 		Plotting the Map Plot: ChatGPT provided guidance in understanding the approach to create and interpret map plots, enhancing the geographical aspect of our analysis.
* 		Highlighting Key Points: ChatGPT aided in highlighting the local maxima and minima an in our line graphs, making our insights more accessible and visually compelling.

* 		Graph Beautification: ChatGPT assisted in adjusting the formatting of float values for years on the X-axis, enhancing the clarity of several graphs.

2. Official documentation of Pandas, Numpy, Seaborn and Matplotlib for further clarification of functions/arguments.



# **Next Steps**

